### ISSUE TITLE
aws_lambda_nodejs: specify outDir when using preCompilation 

### ISSUE THREAD

**Issue by mathieubrd** [aws/aws-cdk#19841]
### Describe the feature

When bundling an AWS Lambda using the `preCompilation` option, JS files are emitted in the same location as the TS files.
When working with a lot of TS files, it's very annoying to get dozen of JS files generated.
We should be able to define an output directory.
Pre-compilation forces the `outDir` option to be `./` which totally overrides the value defined in the `tsconfig.json`.

### Use Case

When doing a `cdk synth` or `cdk build` when the `preCompilation` option is used, it's frustrating to get JS files generated in the same location as the TS files.
I would like to specify an output directory for the JS files but the `outDir` option from the `tsconfig.json` is ignored.

### Proposed Solution

The function `getTsconfigCompilerOptions` force the `outDir` to `./`.
```
const options: Record<string, any> = {
    ...compilerOptions,
    // Overrides
    incremental: false,
    // Intentionally Setting rootDir and outDir, so that the compiled js file always end up next to .ts file.
    rootDir: './',
    outDir: './',
};
```
If the `outDir` option is defined in the `tsconfig.json`, it should be taken into account.
I don't know what would be the consequence for `esbuild` (will be able to find the built JS file?).

### Acknowledgements

- [ ] I may be able to implement this feature request
- [X] This feature might incur a breaking change

### CDK version used

2.20.0

### Environment details (OS name and version, etc.)

macOS 12.2.1

**Comment by corymhall** [comment]
I'm not sure if there was a reason that `outDir` was intentionally set since I don't think esbuild
cares about the that. @hassanazharkhan, @jogold do you know?

**Comment by jogold** [comment]
I think we set `outDir` because this way we can determine where the js handler file will end up after `tsc` and give this to `esbuild`. Otherwise we need to extract it from the tsconfig and write a logic to discover where the compiled handler file ends up.

**Comment by mathieubrd** [comment]
> I think we set `outDir` because this way we can determine where the js handler file will end up after `tsc` and give this to `esbuild`. Otherwise we need to extract it from the tsconfig and write a logic to discover where the compiled handler file ends up.

I think it would be a plus to get this feature.
It's really frustrating to get our workspace polluted by dozen of JS files when using `preCompilation`.
Can we consider this?

I will try to do a PR in a near future.

**Comment by patrickcollings** [comment]
Any update with this?

**Comment by owenbirdsall** [comment]
Yes, I would love to have a fix for this - but if not (and if there really is no workaround) then it would be great to highlight that there is no solution for this problem in the docs, as I just wasted half a day trying to find a solution

**Comment by mathieubrd** [comment]
> > I think we set `outDir` because this way we can determine where the js handler file will end up after `tsc` and give this to `esbuild`. Otherwise we need to extract it from the tsconfig and write a logic to discover where the compiled handler file ends up.
> 
> I think it would be a plus to get this feature. It's really frustrating to get our workspace polluted by dozen of JS files when using `preCompilation`. Can we consider this?
> 
> I will try to do a PR in a near future.

Spoiler alert, I did not

**Comment by andrei-cdl** [comment]
as a intermittent step would it be simple enough to have a clean up step instead? keep track of all generated file and once the build is done and `cdk.out` assets are created for the final bundle to iterate over the project directory and select `d.ts` and `.js` that match the `ts` files.

We don't want to remove valid `.js` or `d.ts` files so this might be simple enough approach to help and can be a setting rather then default.

**Comment by nateq314** [comment]
Bump, it's March 2024 and I'm running into this issue. Any progress?
